concordium-consensus/src/Concordium/GlobalState/Persistent/Trie.hs

```diff
@@ -76,16 +83,86 @@ instance OrdFixedTrieKey Word32
 deriving via Word64 instance OrdFixedTrieKey BakerId
 deriving via Word64 instance OrdFixedTrieKey DelegatorId
 
+-- |A pair of an index and a branch.
+data BranchEntry r = BranchEntry
+    { -- |The branch index.
+      beIndex :: {-# UNPACK #-} !Word8,
+      -- |The branch itself.
+      beBranch :: !r
+    }
+    deriving (Show, Functor, Foldable, Traversable)
+
+-- |Data structure representing the branches of the Trie.
+newtype Branches r = Branches
+    { -- |The branches, represented as an ordered array of pairs of indices and non-null values.
+      theBranches :: Array.Array (BranchEntry r)
+    }
+    deriving (Show, Functor, Foldable, Traversable)
+
+-- |Convert 'Branches' to a list. The list will have length 256.
+branchesToList :: Branches r -> [Nullable r]
+branchesToList = mkl 0 . Foldable.toList . theBranches
+  where
+    mkl n [] = replicate (256 - n) Null
+    mkl n (BranchEntry i v : r) = replicate (fromIntegral i - n) Null ++ Some v : mkl (fromIntegral i + 1) r
+
+-- |Convert a list to 'Branches'. The list MUST have length 256.
+branchesFromList :: [Nullable r] -> Branches r
+branchesFromList = Branches . Array.fromList . mkl 0
+  where
+    mkl _ [] = []
+    mkl !i (Null : r) = mkl (i + 1) r
+    mkl !i (Some v : r) = BranchEntry i v : mkl (i + 1) r
+
+-- |Get the branch at a particular index.
+branchAt :: Branches r -> Word8 -> Nullable r
+branchAt br i =
+    case binarySearch beIndex (V.fromArray (theBranches br)) i of
+        Nothing -> Null
+        Just (BranchEntry _ v) -> Some v
+
+-- |Update the branch at a particular index.
+-- This is strict in the value.
+updateBranch :: Word8 -> Nullable r -> Branches r -> Branches r
+updateBranch i (Some v) = Branches . Array.fromList . updl . Foldable.toList . theBranches
+  where
+    updl [] = [BranchEntry i v]
+    updl l@(p@(BranchEntry k _) : r) = case compare i k of
+        LT -> BranchEntry i v : l
+        EQ -> BranchEntry i v : r
+        GT -> p : updl r
+updateBranch i Null = Branches . Array.fromList . updl . Foldable.toList . theBranches
+  where
+    updl [] = []
+    updl l@(p@(BranchEntry k _) : r) = case compare i k of
+        LT -> l
+        EQ -> r
+        GT -> p : updl r
+
+-- |Construct a 'Branches' with the given entries; all other entries are 'Null'.
+-- This is strict in the provided values.
+makeBranches :: [(Word8, Nullable r)] -> Branches r
+makeBranches = Branches . Array.fromList . collapse . sortOn fst
+  where
+    collapse [] = []
+    collapse [(i, Some v)] = [BranchEntry i v]
+    collapse [(_, Null)] = []
+    collapse ((i1, v1) : t@((i2, _) : _))
+        | i1 == i2 = collapse t
```

Comment:
Why is this special case needed? Why can't we make the assumption that keys are distinct?

Is there also an assumption that the list is ordered on the first component?

Reply:
I have replaced this with a simpler function that is better suited to how this is used. There is no ordering assumption, but we can assume the keys are distinct - and in fact there are only two keys.